frame/contracts/proc-macro/src/lib.rs

```diff
@@ -140,3 +147,407 @@ fn format_default(field: &Ident) -> TokenStream {
+struct HostFn {
+	item: syn::ItemFn,
+	module: String,
+	name: String,
+	returns: HostFnReturn,
+}
+
+enum HostFnReturn {
+	Unit,
+	U32,
+	ReturnCode,
+}
+
+/// Helper trait to convert a host function definition into its wasm signature.
+trait ToWasmSig {
+	fn to_wasm_sig(&self) -> TokenStream;
+}
+
+impl ToWasmSig for HostFn {
+	fn to_wasm_sig(&self) -> TokenStream {
+		let args = self.item.sig.inputs.iter().skip(1).filter_map(|a| match a {
+			syn::FnArg::Typed(pt) => Some(&pt.ty),
+			_ => None,
+		});
+		let returns = match &self.returns {
+			HostFnReturn::U32 => quote! { vec![ <u32>::VALUE_TYPE ] },
+			HostFnReturn::ReturnCode => quote! { vec![ <ReturnCode>::VALUE_TYPE ] },
+			HostFnReturn::Unit => quote! { vec![] },
+		};
+
+		quote! {
+			 wasm_instrument::parity_wasm::elements::FunctionType::new(
+				vec! [ #(<#args>::VALUE_TYPE),* ],
+				#returns,
+			)
+		}
+	}
+}
+
+impl ToTokens for HostFn {
+	fn to_tokens(&self, tokens: &mut TokenStream) {
+		self.item.to_tokens(tokens);
+	}
+}
+
+impl HostFn {
+	pub fn try_from(item: syn::Item) -> syn::Result<Self> {
+		let err = |span, msg| {
+			let msg = format!("Invalid host function definition. {}", msg);
+			syn::Error::new(span, msg)
+		};
+		let msg = "only #[version(<u8>)] or #[unstable] attribute is allowed.";
+		let span = item.span();
+		let item = match item {
+			syn::Item::Fn(i_fn) => Ok(i_fn),
+			_ => Err(err(span, msg)),
+		}?;
+
+		let name = item.sig.ident.to_string();
+		let mut module = "seal0".to_string();
+		let attrs: Vec<&syn::Attribute> =
+			item.attrs.iter().filter(|m| !m.path.is_ident("doc")).collect();
+
+		match attrs.len() {
+			0 => (),
+			1 => {
+				let attr = &attrs[0];
+				let ident = attr.path.get_ident().ok_or(err(span, msg))?.to_string();
+				match ident.as_str() {
+					"version" => {
+						let ver: syn::LitInt = attr.parse_args()?;
+						module = format!(
+							"seal{}",
+							ver.base10_parse::<u8>().map_err(|_| err(span, msg))?
+						);
+					},
+					"unstable" => {
+						module = "__unstable__".to_string();
+					},
+					_ => return Err(err(span, msg)),
+				}
+			},
+			_ => return Err(err(span, msg)),
+		}
+
+		let msg = format!(
+			r#"Should return one of the following: 
+								  - Result<(), TrapReason>,
+								  - Result<ReturnCode, TrapReason>,
+								  - Result<u32, TrapReason>"#
+		);
+
+		let ret_ty = match item.clone().sig.output {
+			syn::ReturnType::Type(_, ty) => Ok(ty.clone()),
+			_ => Err(err(span, &msg)),
+		}?;
+
+		match *ret_ty {
+			syn::Type::Path(tp) => {
+				let result = &tp.path.segments.first().ok_or(err(span, &msg))?;
+				let (id, span) = (result.ident.to_string(), result.ident.span());
+				id.eq(&"Result".to_string()).then_some(()).ok_or(err(span, &msg))?;
+
+				match &result.arguments {
+					syn::PathArguments::AngleBracketed(group) => {
+						group.args.len().eq(&2).then_some(42).ok_or(err(span, &msg))?;
+
+						let arg2 = group.args.last().ok_or(err(span, &msg))?;
```

Comment:
Could possibly simply the args extraction with slice pattern matching e,g, `if let  [arg1, arg2] = &group.args[..] { }`

Reply:
this would add up one more indent lvl :see_no_evil:

Reply:
You could do something like:

```
let (arg1, arg2) = if let [arg1, arg2] = &group.args[..]  {
  Ok((arg1, arrg2))
} else {
  Err("Expected exactly 2 args")
}?;
```

Reply:
would be great, but `&group.args` is [`Punctuated`](https://docs.rs/syn/1.0.99/syn/punctuated/struct.Punctuated.html) and even `.iter().collect()` won't work on it (_the trait `FromIterator<GenericArgument>` is not implemented for `GenericArgument`_)